core/src/Streamly/Internal/Data/Unfold.hs

```diff
@@ -458,6 +462,47 @@ scan (Fold stepF initial extract) (Unfold stepU injectU) =
 
     step Nothing = return Stop
 
+{-# INLINE_NORMAL scanMany #-}
+scanMany :: Monad m => Fold m b c -> Unfold m a b -> Unfold m a c
+scanMany (Fold stepF initial extract) (Unfold stepU injectU) =
+    Unfold step inject
+
+    where
+
+    inject a =  do
+        r <- initial
+        case r of
+            FL.Partial fs -> do
+                x <- injectU a
+                return (Just (fs, x), Nothing)
+            FL.Done _ -> return (Nothing, Nothing)
+
+    {-# INLINE_LATE step #-}
+    step (Just (fs, us), nxt) = do
+        ru <- stepU us
+        case ru of
+            Yield x s -> do
+                rf <- stepF fs x
+                case rf of
+                    FL.Done v -> do
+                        if isNothing nxt
+                        then return $ Yield v (Just (fs, s), Just v)
+                        else do
+                            r <- initial
+                            return $ case r of
+                                FL.Partial fs1 ->
+                                    Skip (Just (fs1, us), Nothing)
+                                FL.Done _ ->
+                                    Skip (Nothing, Nothing)
```

Comment:
This is unnecessarily complicated, why not just say return `Stop` here? The `Maybe` is not required on the first element of the tuple.

Reply:
Better would be to just change `scan` to `scanWith` and pass a flag to it to stop or restart the scan. Then use the same common function of scan and scanMany.